include/deal.II/multigrid/mg_transfer_interface.h

```diff
@@ -0,0 +1,424 @@
+};
+
+
+
+/**
+ * Implementation of the MGTransferBase interface for which the transfer
+ * operations is implemented in a matrix-free way based on the interpolation
+ * matrices of the underlying finite element. This requires considerably less
+ * memory than MGTransferPrebuilt and can also be considerably faster than that
+ * variant. In contrast to MGTransferMatrixFree, the user can provide separate
+ * transfer operators between each level.
+ *
+ * This class currently only works for tensor-product finite elements based on
+ * FE_Q and FE_DGQ elements. Systems involving multiple components of
+ * one of these element, as well as, systems with different elements or other
+ * elements are currently not implemented.
+ */
+template <typename MatrixType>
+class MGTransferMatrixFreeNew
+  : public dealii::MGTransferBase<dealii::LinearAlgebra::distributed::Vector<
+      typename MatrixType::value_type>>
+{
+public:
+  /**
+   * Dimension.
+   */
+  static const int dim = MatrixType::dim;
+
+  /**
+   * Value type.
+   */
+  using Number = typename MatrixType::value_type;
+
+  /**
+   * Constructor.
+   */
+  MGTransferMatrixFreeNew(const MGLevelObject<MatrixType> &           matrices,
+                          const MGLevelObject<Transfer<dim, Number>> &transfer);
+
+  /**
+   * Perform prolongation.
+   */
+  void
+  prolongate(
+    const unsigned int                                        to_level,
+    dealii::LinearAlgebra::distributed::Vector<Number> &      dst,
+    const dealii::LinearAlgebra::distributed::Vector<Number> &src) const;
+
+  /**
+   * Perform restriction.
+   */
+  virtual void
+  restrict_and_add(
+    const unsigned int                                        from_level,
+    dealii::LinearAlgebra::distributed::Vector<Number> &      dst,
+    const dealii::LinearAlgebra::distributed::Vector<Number> &src) const;
+
+  /**
+   * Initialize internal vectors and copy @p src vector to the finest
+   * multigrid level.
+   *
+   * @note DoFHandler is not needed here, but is required by the interface.
+   */
+  template <class InVector, int spacedim>
+  void
+  copy_to_mg(
+    const DoFHandler<dim, spacedim> &dof_handler,
+    MGLevelObject<dealii::LinearAlgebra::distributed::Vector<Number>> &dst,
+    const InVector &src) const;
+
+  /**
+   * Initialize internal vectors and copy the values on the finest
+   * multigrid level to @p dst vector.
+   *
+   * @note DoFHandler is not needed here, but is required by the interface.
+   */
+  template <class OutVector, int spacedim>
+  void
+  copy_from_mg(
+    const DoFHandler<dim, spacedim> &dof_handler,
+    OutVector &                      dst,
+    const MGLevelObject<dealii::LinearAlgebra::distributed::Vector<Number>>
+      &src) const;
+
+private:
+  const MGLevelObject<MatrixType> &           matrices;
+  const MGLevelObject<Transfer<dim, Number>> &transfer;
+};
+
+
+
+/**
+ * Class for repartitioning data living on the same triangulation, which
```

Comment:
triangulation or dofhandler? Or is this more generic? Are these vectors the same size?

This is probably unrelated to multigrid and could go somewhere else (vectortools?).

Reply:
> This is probably unrelated to multigrid and could go somewhere else (vectortools?).

But for now I would like to keep in this PR since it uses the same infrastructure as the global-coarsening algorithm is using.